### ISSUE TITLE
Case insensitive string comparison operators: eqi, nei, lti, gti, lei, gei

### ISSUE THREAD

**Issue by epa** [Perl/perl5#17523]
It would be useful to have infix operators for case-insensitive string comparison.
`$foo eqi $bar`
would be equivalent to
`fc($foo) eq fc($bar)`
and so on.

Yes, it's just syntactic sugar really, but Perl hardly takes an austere approach to these things.  In my view a case-insensitive string comparison falls under "easy things should be easy".

**Comment by demerphq** [comment]
On Mon, 3 Feb 2020 at 15:59, Ed Avis <notifications@github.com> wrote:

> It would be useful to have infix operators for case-insensitive string
> comparison.
> $foo eqi $bar
> would be equivalent to
> fc($foo) eq fc($bar)
> and so on.
>
fc() ? Does that mean "fold-case" ? Did I miss us adding that?

$ ./perl -le'print $]; print fc("foo")'
5.031009
Undefined subroutine &main::fc called at -e line 1.

If you do mean foldcase() I think get why we might use unicode fold casing
rules, but i think some folks might find it odd that for some strings

lc($a) eq lc($b) != uc($a) eq uc($b) != fc($a) eq fc($b)

which iirc is quite possible with certain unicode code points.

> Yes, it's just syntactic sugar really, but Perl hardly takes an austere
> approach to these things. In my view a case-insensitive string comparison
> falls under "easy things should be easy".
>
I guess they would have the same precedence as their case sensitive
equivalents?

Yves


-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by epa** [comment]
Yes, `fc` is fold-case and was added in 5.16: https://perldoc.perl.org/functions/fc.html
It exists because you can get subtle bugs (Turkish, etc) if you use `lc` to normalize two strings for comparison and more so if you use `uc`.  Which is another reason to provide a builtin operator for case-insensitive comparison: it sounds simple, but is actually quite hard to get right.

Yes, I envisage `eqi` being a drop-in substitute for `eq`, and so on.

**Comment by Grinnz** [comment]
fc() is available with https://perldoc.pl/feature#The-%27fc%27-feature

**Comment by demerphq** [comment]
On Mon, 3 Feb 2020 at 16:18, Ed Avis <notifications@github.com> wrote:

> Yes, fc is fold-case and was added in 5.16:
> https://perldoc.perl.org/functions/fc.html
>

I missed that, I guess I have to enable it with a use feature or something
right? (Dont answer, ill read the docs ;-)


>
> It exists because you can get subtle bugs (Turkish, etc) if you use lc to
> normalize two strings for comparison and more so if you use uc.
>
Yeah, im familiar with fold-casing. Just one note, afaik Turkish is (the?)
one language that Unicode doesn't solve with fold-casing.  See

https://en.wikipedia.org/wiki/Dotted_and_dotless_I

"Most Unicode software uppercases *ı* to *I* and lowercases *İ* to *i*,
but, unless specifically configured for Turkish, it lowercases *I* to *i* and
uppercases *i* to *I*. Thus uppercasing then lowercasing, or vice versa,
changes the letters."


> Which is another reason to provide a builtin operator for case-insensitive
> comparison: it sounds simple, but is actually quite hard to get right.
>
Totally get you, I warm to the idea actually, even if I can imagine others
wont.

yves
ps: but dont expect it to solve Turkish casing problems, that is only
possible if you know you are dealing with Turkish. :-)

**Comment by Leont** [comment]
I don't think this is a good idea, for the simple reason that there are too many ways to compare strings.
Folding case really is just the start, there are also various normalization forms (NFC, NFD) and collations (unicode and locale, e.g. Turkish).

**Comment by demerphq** [comment]
On Mon, 3 Feb 2020 at 16:45, Leon Timmermans <notifications@github.com>
wrote:

> I don't think this is a good idea, for the simple reason that there are
> too many ways to compare strings.
> Folding case really is just the start, there are also various
> normalization forms (NFC, NFD) and collations (unicode and locale).
>
Since we have precedent for using foldcasing rules (and not worrying about
NFC/NFD) in the regex engine under /i this argument doesn't seem to me to
be very compelling. FWIW.

cheers,
Yves

**Comment by epa** [comment]
Another argument for an `eqi` operator is that it could be implemented more efficiently than `fc($x) eq fc($y)`.  If the first character differs you don't need to continue fold-casing the rest of the string, and so on.

**Comment by leonerd** [comment]
The peephole optimiser is good at spotting patterns like that. It could easily be spelled `fc($a) eq fc($b)` but compiled to some other efficient folding-based operator internally.

E.g. compare also such patterns as `foreach reverse @array` - this doesn't form a temporary array in the reversed order, but simply iterates the given array backwards.

This has the advantage of making the runtime performance of existing syntax faster, while not breaking code so a given program still runs correctly on older perls that simply lack that optimisation

**Comment by Leont** [comment]
> The peephole optimiser is good at spotting patterns like that. It could easily be spelled fc($a) eq fc($b) but compiled to some other efficient folding-based operator internally.

Exactly

**Comment by epa** [comment]
Dear perl5 developers, could I ask you to have another look at this feature request?

I believe that case-insensitive string comparison operators like `eqi`, `lti`, `cmpi` and so on will be a useful addition to the language and that precedent for case-folding rules is already given by the `fc` operator and the `/i` flag on regexps.

**Comment by leonerd** [comment]
I'm inclined to disagree.

I have plans to introduce an `equ` operator which would care about undef vs defined strings. I.e `undef equ undef` is true, `undef equ ""` is false with no warning, and otherwise would behave like `eq` for two defined strings.

With that one added, would we then add two *more* operators for `eqi` and `equi` (or would it be spelled `eqiu`)?

Given that your proposed `eqi` can be so easily spelled using the `fc()` function, whereas my `equ` cannot, I would vote if only one is to be added, `equ` be it.

**Comment by jgamble** [comment]
If that's the case, then is there any plan to add fc detection as mentioned [above](https://github.com/Perl/perl5/issues/17523#issuecomment-581994444) to the peephole optimizer?

**Comment by Leont** [comment]
> If that's the case, then is there any plan to add fc detection as mentioned above to the peephole optimizer?

We're (mostly) a volunteer based organization. It will happen if someone volunteers themselves.

All the pieces already exists (in particular there's `foldEQ_utf8_flags`), someone just has to tie them together (I kind of suspect Deparse is the hardest bit, TBH).

**Comment by jgamble** [comment]
I'm aware of that, but if there hasn't even been a PR submitted after saying "it can be done just like reverse array", well, I feel that that at least should be a step. And to forestall the inevitable comment no, at the moment I'm not ready to even write a PR, such is my unfamiliarity with the internals.
